src/picongpu/include/particles/ionization/byField/Keldysh/Keldysh_Impl.hpp

```diff
@@ -0,0 +1,221 @@
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <boost/type_traits/integral_constant.hpp>
+
+#include "simulation_defines.hpp"
+#include "traits/Resolve.hpp"
+#include "traits/UsesRNG.hpp"
+#include "mappings/kernel/AreaMapping.hpp"
+
+#include "fields/FieldB.hpp"
+#include "fields/FieldE.hpp"
+
+#include "particles/ionization/byField/Keldysh/Keldysh.def"
+#include "particles/ionization/byField/Keldysh/AlgorithmKeldysh.hpp"
+#include "particles/ionization/ionization.hpp"
+
+#include "compileTime/conversion/TypeToPointerPair.hpp"
+#include "memory/boxes/DataBox.hpp"
+
+#include "particles/ionization/ionizationMethods.hpp"
+
+#include "random/methods/XorMin.hpp"
+#include "random/distributions/Uniform.hpp"
+#include "random/RNGProvider.hpp"
+
+namespace picongpu
+{
+namespace traits
+{
+    /** specialization of the UsesRNG trait
+     * --> ionization module uses random number generation
+     */
+    template<typename T_IonizationAlgorithm, typename T_DestSpecies, typename T_SrcSpecies>
+    struct UsesRNG<particles::ionization::Keldysh_Impl<T_IonizationAlgorithm, T_DestSpecies, T_SrcSpecies> > :
+    public boost::true_type
+    {
+    };
+} // namespace traits
+
+namespace particles
+{
+namespace ionization
+{
+
+    /** \struct Keldysh_Impl
+     *
+     * \brief Ammosov-Delone-Krainov
+     *        Tunneling ionization for hydrogenlike atoms
+     *
+     * \tparam T_DestSpecies electron species to be created
+     * \tparam T_SrcSpecies particle species that is ionized
+     */
+    template<typename T_IonizationAlgorithm, typename T_DestSpecies, typename T_SrcSpecies>
+    struct Keldysh_Impl
+    {
+
+        typedef T_DestSpecies DestSpecies;
+        typedef T_SrcSpecies  SrcSpecies;
+
+        typedef typename SrcSpecies::FrameType FrameType;
+
+        /* specify field to particle interpolation scheme */
+        typedef typename PMacc::traits::Resolve<
+            typename GetFlagType<FrameType,interpolation<> >::type
+        >::type Field2ParticleInterpolation;
+
+        /* margins around the supercell for the interpolation of the field on the cells */
+        typedef typename GetMargin<Field2ParticleInterpolation>::LowerMargin LowerMargin;
+        typedef typename GetMargin<Field2ParticleInterpolation>::UpperMargin UpperMargin;
+
+        /* relevant area of a block */
+        typedef SuperCellDescription<
+            typename MappingDesc::SuperCellSize,
+            LowerMargin,
+            UpperMargin
+            > BlockArea;
+
+        BlockArea BlockDescription;
+
+        private:
+
+            /* define ionization ALGORITHM (calculation) for ionization MODEL */
+            typedef T_IonizationAlgorithm IonizationAlgorithm;
+
+            /* random number generator */
+            typedef PMacc::random::RNGProvider<simDim, PMacc::random::methods::XorMin> RNGFactory;
+            typedef PMacc::random::distributions::Uniform<float> Distribution;
```

Comment:
copied from `particles/synchrotronPhotons/PhotonCreator.hpp` and `particles/ionization/byField/ADK/ADK_Impl.hpp`